collector/key_users_linux.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build linux && !nokey_users
+// +build linux,!nokey_users
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	keyUsersSubsystem = "keys"
+)
+
+var (
+	// See https://man7.org/linux/man-pages/man7/keyrings.7.html
+	keyUsersLineRegex = regexp.MustCompile(`^\s*(?P<uid>\d+):\s+(?P<usage>\d+)\s+(?P<nkeys>\d+)/(?P<nikeys>\d+)\s+(?P<qnkeys>\d+)/(?P<maxkeys>\d+)\s+(?P<qnbytes>\d+)/(?P<maxbytes>\d+)$`)
+)
+
+type keyUsersCollector struct {
+	logger log.Logger
+}
+
+type keyUsersEntry struct {
+	uid  int64
+	name string
+}
+
+func init() {
+	registerCollector("key_users", defaultDisabled, NewKeyUsersCollector)
+}
+
+func NewKeyUsersCollector(logger log.Logger) (Collector, error) {
+	return &keyUsersCollector{logger}, nil
+}
+
+func (c *keyUsersCollector) Update(ch chan<- prometheus.Metric) error {
+	keyUsers, err := c.getKeyUsers()
+	if err != nil {
+		return fmt.Errorf("couldn't get key-users: %w", err)
+	}
+	level.Debug(c.logger).Log("msg", "Set node_keys", "keyUsers", keyUsers)
+	for k, v := range keyUsers {
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, keyUsersSubsystem, k.name),
+				fmt.Sprintf("Key utilization field %s.", k.name),
+				[]string{"uid"}, nil,
+			),
+			prometheus.GaugeValue, v, strconv.FormatInt(int64(k.uid), 10),
+		)
+	}
+	return nil
+}
+
+func (c *keyUsersCollector) getKeyUsers() (map[keyUsersEntry]float64, error) {
+	file, err := os.Open(procFilePath("key-users"))
```

Comment:
Parsing files in `/proc` is something we do in https://github.com/prometheus/procfs first. Please implement parsing there before adding this to the exporter.